### PR TITLE
Add support for gzip decoding to blackholes

### PR DESCRIPTION
**Pull request by jszwedko**
Best reviewed with whitespace changes ignored.

Signed-off-by: Jesse Szwedko <jesse.szwedko@gmail.com>

### CODE REVIEW

**Comment by jszwedko** [comment]
@blt thanks for the prompt! I integrated it into refactored code base if you want to take a 👀

File: src/codec.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
+use bytes::{Buf, Bytes};
+use flate2::read::{MultiGzDecoder, ZlibDecoder};
+use hyper::{Body, StatusCode};
+use std::io::Read;
+
+pub fn decode(
+    content_type: Option<&hyper::header::HeaderValue>,
+    mut body: Bytes,
+) -> Result<Bytes, hyper::Response<hyper::Body>> {
```
Comment (jszwedko) [inline src/codec.rs:9]: I briefly thought about trying to abstract this some more and move the hyper bits but figured I'd defer until we actually needed a reason to abstract.

File: src/codec.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    if let Some(content_type) = content_type {
+        let content_type = String::from_utf8_lossy(content_type.as_bytes());
```
[the remark this answers is not in the document]
Comment (jszwedko) [reply src/codec.rs:11]: Yeah, I thought about that, but wanted to be able to use `str::trim`. There didn't seem to be a ready way to trim a byte slice so I just opted to coerce it to a string (as Vector does). I'm happy to introduce a function to trim a byte slice if we want though.

[the remark this answers is not in the document]
Comment (jszwedko) [reply src/codec.rs:11]: I borrowed this implementation from https://github.com/vectordotdev/vector/blob/aea69e02cf4992ec9b611eb3b2ebce3f6db1da37/src/sources/util/http/encoding.rs#L11-L45

File: src/codec.rs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+        for encoding in content_type.rsplit(',').map(str::trim) {
+            body = match encoding {
+                "identity" => body,
```
[the remark this answers is not in the document]
Comment (jszwedko) [reply src/codec.rs:15]: Good call. I copied this from https://github.com/vectordotdev/vector/blob/aea69e02cf4992ec9b611eb3b2ebce3f6db1da37/src/sources/util/http/encoding.rs#L11-L45 but you are right that it should be case-insensitive per https://httpwg.org/specs/rfc7231.html#content.codings

File: src/codec.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (blt) [reply src/codec.rs]: Works for me.

**Review by blt — APPROVED** [review]

**Pull request merged**